relay-server/src/actors/store.rs

```diff
@@ -199,51 +199,122 @@ impl StoreService {
+            if matches!(event_item.ty(), ItemType::Transaction) {
+                self.extract_kafka_attachments(
+                    &mut kafkamsgs,
+                    attachments.clone(),
+                    &event_id,
+                    &scoping,
+                );
+                attachments.clear();
+            }
+
             let event_message = KafkaMessage::Event(EventKafkaMessage {
                 payload: event_item.payload(),
                 start_time: UnixTimestamp::from_instant(start_time).as_secs(),
-                event_id: event_id.ok_or(StoreError::NoEventId)?,
+                event_id: match event_id {
+                    Some(event_id) => event_id,
+                    None => return kafkamsgs,
+                },
                 project_id: scoping.project_id,
-                remote_addr: envelope.meta().client_addr().map(|addr| addr.to_string()),
-                attachments,
+                remote_addr,
+                attachments: attachments.clone(),
             });
 
-            self.produce(topic, scoping.organization_id, event_message)?;
-            metric!(
-                counter(RelayCounters::ProcessingMessageProduced) += 1,
-                event_type = &event_item.ty().to_string()
+            kafkamsgs.push(event_message);
+        } else {
+            self.extract_kafka_attachments(
+                &mut kafkamsgs,
+                attachments.clone(),
+                &event_id,
+                &scoping,
             );
-        } else if !attachments.is_empty() {
+        }
+
+        kafkamsgs
+    }
+
+    fn extract_kafka_attachments(
+        &self,
+        kafkamsgs: &mut Vec<KafkaMessage>,
+        attachments: Vec<ChunkedAttachment>,
+        event_id: &Option<EventId>,
+        scoping: &Scoping,
+    ) {
+        if !attachments.is_empty() {
             relay_log::trace!("Sending individual attachments of envelope to kafka");
+
             for attachment in attachments {
                 let attachment_message = KafkaMessage::Attachment(AttachmentKafkaMessage {
-                    event_id: event_id.ok_or(StoreError::NoEventId)?,
+                    event_id: match event_id {
+                        Some(event_id) => *event_id,
+                        None => continue,
+                    },
                     project_id: scoping.project_id,
                     attachment,
                 });
 
-                self.produce(topic, scoping.organization_id, attachment_message)?;
-                metric!(
-                    counter(RelayCounters::ProcessingMessageProduced) += 1,
-                    event_type = "attachment"
-                );
+                kafkamsgs.push(attachment_message);
             }
         }
-
-        Ok(())
     }
 
     fn produce(
         &self,
         topic: KafkaTopic,
         organization_id: u64,
-        message: KafkaMessage,
+        message: &KafkaMessage,
```

Comment:
this created a lot of diffs. I can start a new PR that only changes the signature to make this one more clear.